### PR TITLE
IBMCloud: BYON Enablement - InstallConfig

### PR DESCRIPTION
**Pull request by cjschaef**
Setup enablement for Bring Your Own Network support on IBM Cloud
using IPI. This portion focuses on adding support for the
required values in the InstallConfig, and performing simple
validation of those resources provided.

### CODE REVIEW

**Comment by cjschaef** [comment]
This requires https://github.com/openshift/installer/pull/6046 changes

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 1]
```diff
@@ -377,6 +397,34 @@ func (c *Client) GetVPC(ctx context.Context, vpcID string) (*vpcv1.VPC, error) {
...
+	defer cancel()
+
+	regions, err := c.getVPCRegions(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	allVPCs := []vpcv1.VPC{}
+	for _, region := range regions {
+		err := c.setVPCServiceURLForRegion(ctx, *region.Name)
```
Comment (cjschaef) [inline pkg/asset/installconfig/ibmcloud/client.go:412]: I will need to rebase this PR in order to fix this call now that the necessary PR has merged,
https://github.com/openshift/installer/pull/6032
```
pkg/asset/installconfig/ibmcloud/client.go:413:11: c.setVPCServiceURLForRegion undefined (type *Client has no field or method setVPCServiceURLForRegion, but does have SetVPCServiceURLForRegion)
```

**Comment by cjschaef** [comment]
I believe I resolved the remaining issues after the rebase off #6050 

```
# bin/openshift-install version
bin/openshift-install unreleased-master-6173-g273656b43019522d2eeebcd5f367c50e473a7f4b
built from commit 273656b43019522d2eeebcd5f367c50e473a7f4b
release image registry.ci.openshift.org/origin/release:4.11
release architecture amd64

# yq '.platform.ibmcloud' cluster-deploys/cjs-test-107/install-config.yaml 
region: us-east
resourceGroupName: existing-network-1
vpcName: existing-network-1-vpc
controlPlaneSubnets:
  - existing-network-1-us-east-1
  - existing-network-1-us-east-2
  - existing-network-1-us-east-3
computeSubnets:
  - existing-compute-network-1-us-east-1
  - existing-compute-network-1-us-east-2
  - existing-compute-network-1-us-east-3

# bin/openshift-install create manifests --dir cluster-deploys/cjs-test-107
INFO Consuming Install Config from target directory 
INFO Manifests created in: cluster-deploys/cjs-test-107/manifests and cluster-deploys/cjs-test-107/openshift
```

/unhold

**Review by rvanderp3 — COMMENTED** [review]
found a few nits, otherwise lgtm

**Comment by cjschaef** [comment]
Sorry, will get to these soon, have been on other non-IPI duties

File: pkg/asset/installconfig/ibmcloud/validation.go [hunk 2]
```diff
@@ -218,6 +222,89 @@ func validateResourceGroup(client API, ic *types.InstallConfig, path *field.Path
...
+					allErrs = append(allErrs, field.Invalid(path.Child("controlPlaneSubnets"), controlPlaneSubnet, fmt.Sprintf("controlPlaneSubnets contains subnet: %s, not found in expected vpcID: %s", controlPlaneSubnet, vpcID)))
+				}
+				if *subnet.ResourceGroup.ID != ic.IBMCloud.ResourceGroupName && *subnet.ResourceGroup.Name != ic.IBMCloud.ResourceGroupName {
+					allErrs = append(allErrs, field.Invalid(path.Child("controlPlaneSubnets"), controlPlaneSubnet, fmt.Sprintf("controlPlaneSubnets contains subnet: %s, not found in expected resourceGroupName: %s", controlPlaneSubnet, ic.IBMCloud.ResourceGroupName)))
+				}
+			}
+		}
+	}
+
+	if ic.IBMCloud.ComputeSubnets == nil || len(ic.IBMCloud.ComputeSubnets) == 0 {
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/validation.go:283]: Yeah, same as above, but for `Compute`.
https://github.com/openshift/installer/pull/6050/files#diff-da80c878bba406f82e10ab78858cbaff8bb20a72f7aed6ce10051c4ab550441fR50

I can remove the `nil` check

File: pkg/asset/installconfig/ibmcloud/validation.go [hunk 3]
```diff
@@ -218,6 +222,89 @@ func validateResourceGroup(client API, ic *types.InstallConfig, path *field.Path
...
+		return append(allErrs, field.InternalError(path.Child("vpcName"), err))
+	}
+
+	found := false
+	for _, vpc := range vpcs {
+		if *vpc.Name == ic.IBMCloud.VPCName {
+			if *vpc.ResourceGroup.ID != ic.IBMCloud.ResourceGroupName && *vpc.ResourceGroup.Name != ic.IBMCloud.ResourceGroupName {
+				return append(allErrs, field.Invalid(path.Child("vpcName"), ic.IBMCloud.VPCName, fmt.Sprintf("vpc is not in provided ResourceGroup: %s", ic.IBMCloud.ResourceGroupName)))
+			}
+			found = true
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/validation.go:247]: Yep, let me add that post `validateExistingSubnets`

[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/validation.go:247]: added

File: pkg/asset/installconfig/ibmcloud/validation.go [hunk 4]
```diff
@@ -218,6 +222,89 @@ func validateResourceGroup(client API, ic *types.InstallConfig, path *field.Path
...
+	if !found {
+		allErrs = append(allErrs, field.NotFound(path.Child("vpcName"), ic.IBMCloud.VPCName))
+	}
+	return allErrs
+}
+
+func validateExistingSubnets(client API, ic *types.InstallConfig, path *field.Path, vpcID string) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if ic.IBMCloud.ControlPlaneSubnets == nil || len(ic.IBMCloud.ControlPlaneSubnets) == 0 {
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/validation.go]: Fixed

File: pkg/asset/installconfig/ibmcloud/validation.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/validation.go]: Fixed

**Comment by cjschaef** [comment]
/rettitle CORS-2263: IBMCloud: BYON Enablement - InstallConfig

**Pull request merged**